standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreChecker.java

```diff
@@ -303,56 +308,103 @@ void checkTable(Table table, PartitionIterable parts, byte[] filterExp, CheckRes
+            new ArrayBlockingQueue<>(threadCount),
+            new ThreadPoolExecutor.CallerRunsPolicy());
+
+    try {
+      Queue<Future<String>> futures = new LinkedList<>();
+      // check that the partition folders exist on disk
+      for (Partition partition : parts) {
+        if (partition == null) {
+          // most likely the user specified an invalid partition
+          continue;
         }
-      }
+        Path[] partPath = {getDataLocation(table, partition)};
+        if (partPath[0] == null) {
+          continue;
+        }
+        futures.add(pool.submit(new Callable() {
+          @Override
+          public Object call() throws Exception {
+            fs[0] = partPath[0].getFileSystem(conf);
+            CheckResult.PartitionResult prFromMetastore = new CheckResult.PartitionResult();
+            prFromMetastore.setPartitionName(getPartitionName(table, partition));
+            prFromMetastore.setTableName(partition.getTableName());
+            if (!fs[0].exists(partPath[0])) {
+              synchronized (result) {
+                result.getPartitionsNotOnFs().add(prFromMetastore);
+              }
+            } else {
+              synchronized (result) {
+                result.getCorrectPartitions().add(prFromMetastore);
+              }
+            }
+
+            if (partitionExpirySeconds > 0) {
+              long currentEpochSecs = Instant.now().getEpochSecond();
+              long createdTime = partition.getCreateTime();
+              long partitionAgeSeconds = currentEpochSecs - createdTime;
+              if (partitionAgeSeconds > partitionExpirySeconds) {
+                CheckResult.PartitionResult pr = new CheckResult.PartitionResult();
+                pr.setPartitionName(getPartitionName(table, partition));
+                pr.setTableName(partition.getTableName());
+                synchronized (result) {
+                  result.getExpiredPartitions().add(pr);
+                }
+                if (LOG.isDebugEnabled()) {
+                  LOG.debug("{}.{}.{}.{} expired. createdAt: {} current: {} age: {}s expiry: {}s", partition.getCatName(),
+                      partition.getDbName(), partition.getTableName(), pr.getPartitionName(), createdTime, currentEpochSecs,
+                      partitionAgeSeconds, partitionExpirySeconds);
+                }
+              }
+            }
 
-      for (int i = 0; i < getPartitionSpec(table, partition).size(); i++) {
-        Path qualifiedPath = partPath.makeQualified(fs);
-        partPaths.add(qualifiedPath);
-        partPath = partPath.getParent();
+            for (int i = 0; i < getPartitionSpec(table, partition).size(); i++) {
+              final Path qualifiedPath = partPath[0].makeQualified(fs[0]);
+              synchronized (partPaths) {
+                partPaths.add(qualifiedPath);
+              }
+              partPath[0] = partPath[0].getParent();
+            }
+            return "CheckTable task complete.";
```

Comment:
Maybe add something related to the path/partition/table which could be useful in the later debug message. This message is really useless for anything, other than counting the successful checks 😄